File: taskhero/src/tasks/list.rs
// Copyright 2018 Mathew Robinson <chasinglogic@gmail.com>. All rights reserved.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

use super::task::Task;
use serde_json;

use std::fmt;
use std::fmt::Display;
use std::iter;
use std::slice;

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
pub struct List {
    pub tasks: Vec<Task>,
}

impl IntoIterator for List {
    type Item = Task;
    type IntoIter = ::std::vec::IntoIter<Task>;

    fn into_iter(self) -> Self::IntoIter {
        self.tasks.into_iter()
    }
}

impl Display for List {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match serde_json::to_string_pretty(self) {
            Ok(jsn) => write!(f, "{}", jsn),
            Err(_) => write!(f, "ERROR: Unable to serialize list"),
        }
    }
}

impl List {
    pub fn new(tasks: Vec<Task>) -> List {
        let mut list = List { tasks: tasks };

        list.tasks.sort();
        list
    }

    /// Return a reference to the "current" task.
    pub fn current<'a>(&'a mut self) -> Option<&'a mut Task> {
        let mut ind = 0;
        for (task_id, task) in self.enumerate() {
            if !task.completed() {
                ind = task_id;
                break;
            }
        }

        self.find_by_ind(ind)
    }

    /// Add a task to the List, will sort after adding.
    pub fn add(&mut self, task: Task) {
        self.tasks.push(task);
        self.tasks.sort();
    }

    /// Add multiple tasks to the List, this is more efficient than calling add multiple times
    /// since only one sort is performed. It will empty the given vector.
    pub fn add_multiple(&mut self, tasks: &mut Vec<Task>) {
        self.tasks.append(tasks);
        self.tasks.sort();
    }

    /// Return a reference to the task at the given ID / index.
    pub fn find_by_ind<'a>(&'a mut self, id: usize) -> Option<&'a mut Task> {
        if self.tasks.len() < id {
            return None;
        }

        let t: &'a mut Task = &mut self.tasks[id];
        Some(t)
    }

    /// Return a reference to the first task with the given title.
    pub fn find_by_title<'a>(&'a mut self, title: &str) -> Option<&'a mut Task> {
        let mut ind = None;
        for (i, task) in self.enumerate() {
            if task.title == title {
                ind = Some(i);
                break;
            }
        }

        match ind {
            Some(i) => self.find_by_ind(i),
            None => None,
        }
    }

    /// Return an enumerated iterator over the tasks in this list.
    pub fn enumerate(&self) -> iter::Enumerate<slice::Iter<Task>> {
        self.tasks.iter().enumerate()
    }

    pub fn to_json(&self) -> Result<String, serde_json::Error> {
        serde_json::to_string_pretty(self)
    }

    pub fn context(&self, context: &str) -> List {
        List::new(
            self.tasks
                .iter()
                .filter(|x| &x.context == context)
                .map(|x| x.clone())
                .collect(),
        )
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use chrono::prelude::*;
    use chrono::Duration;

    #[test]
    fn test_simple_task_ordering() {
        let list = List::new(vec![
            Task::new("test1").with_priority(1),
            Task::new("test3").with_priority(3),
            Task::new("test2").with_priority(2),
            Task::new("test0"),
        ]);

        let mut priority = 3;
        for task in list {
            assert_eq!(task.priority, priority);
            priority = priority - 1;
        }
    }

    #[test]
    fn test_multi_day_task_ordering() {
        let mut yesterday = Task::new("test2").with_priority(2);
        yesterday.created_date = Local::now() - Duration::days(1);

        let tasks = vec![
            Task::new("test1").with_priority(1),
            yesterday,
            Task::new("test3").with_priority(3),
            Task::new("test0"),
        ];

        let list = List::new(tasks.clone());

        let mut iter = list.into_iter();
        assert_eq!(iter.next().unwrap(), tasks[1]);
        assert_eq!(iter.next().unwrap(), tasks[2]);
        assert_eq!(iter.next().unwrap(), tasks[0]);
        assert_eq!(iter.next().unwrap(), tasks[3]);
    }
}

File: task/list.go
// Copyright 2018 Mathew Robinson <chasinglogic@gmail.com>. All rights reserved.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

package task

import (
	"errors"
	"sort"
	"strconv"
	"strings"
	"time"

	"github.com/chasinglogic/taskforge/ql/ast"
	"github.com/chasinglogic/taskforge/ql/token"
)

// ErrNotFound is returned by a List when a Task matching the given ID does not
// exist
var ErrNotFound = errors.New("no task with that id exists")

// List is implemented by any struct that can maintain tasks
type List interface {
	// Evaluate the AST and return a List of matching results
	Search(ast ast.AST) ([]Task, error)
	// Add a task to the List
	Add(Task) error
	// Add multiple tasks to the List, should be more efficient resource
	// utilization.
	AddMultiple(task []Task) error
	// Return a slice of Tasks in this List
	Slice() []Task
	// Find a task by ID
	FindByID(id string) (Task, error)
	// Return the current task, meaning the oldest uncompleted task in the List
	Current() (Task, error)

	// Complete a task by id
	Complete(id string) error
	// Update a task in the list, finding the original by the ID of the given task
	Update(Task) error
	// Add note to a task by ID
	AddNote(id string, note Note) error
}

// MemoryList implements List for a slice of Tasks
type MemoryList []Task

// TODO: Would counting matching tasks first be faster since we
// frontload allocation?
func (ml *MemoryList) findWhere(f func(t Task) bool) []Task {
	new := []Task{}

	for _, t := range *ml {
		if f(t) {
			new = append(new, t)
		}
	}

	return new
}

func (ml *MemoryList) sort() {
	list := *ml

	sort.Slice(list, func(i, j int) bool {
		return (list[i].Priority > list[j].Priority) ||
			(list[i].Priority == list[j].Priority && list[i].CreatedDate.Before(list[j].CreatedDate))
	})

	ml = &list
}

// Add adds a task to this list
func (ml *MemoryList) Add(task Task) error {
	*ml = append(*ml, task)
	return nil
}

// AddMultiple adds multiple tasks to this list
func (ml *MemoryList) AddMultiple(tasks []Task) error {
	*ml = append(*ml, tasks...)
	return nil
}

// Slice turns the list into a slice of tasks
func (ml *MemoryList) Slice() []Task {
	return []Task(*ml)
}

// FindByID will return a pointer to the task indicated by ID, returns
// ErrNotFound if no task with that ID exists
func (ml *MemoryList) FindByID(id string) (Task, error) {
	for _, t := range *ml {
		if t.ID == id {
			return t, nil
		}
	}

	return Task{}, ErrNotFound
}

// Current will return the first task which is not completed
func (ml *MemoryList) Current() (Task, error) {
	ml.sort()

	for _, t := range *ml {
		if t.CompletedDate.IsZero() {
			return t, nil
		}
	}

	return Task{}, ErrNotFound
}

// Complete will complete the task indicated by id
func (ml *MemoryList) Complete(id string) error {
	t, err := ml.FindByID(id)
	if err != nil {
		return err
	}

	t.CompletedDate = time.Now()
	return ml.Update(t)
}

// Update will update the task indicated by the ID of the provided Task
func (ml *MemoryList) Update(other Task) error {
	list := *ml

	for i := range list {
		if list[i].ID == other.ID {
			list[i].Context = other.Context
			list[i].Priority = other.Priority
			list[i].Body = other.Body
			list[i].Title = other.Title
			list[i].CompletedDate = other.CompletedDate

			if len(other.Notes) > len(list[i].Notes) {
				list[i].Notes = other.Notes
			}

			ml = &list
			return nil
		}
	}

	return ErrNotFound
}

// AddNote will add the note to the task indicated by the ID
func (ml *MemoryList) AddNote(id string, note Note) error {
	t, err := ml.FindByID(id)
	if err != nil {
		return err
	}

	t.Notes = append(t.Notes, note)
	return ml.Update(t)
}

// Search will construct a function to find matching tasks based on the given
// ast.AST
func (ml *MemoryList) Search(tree ast.AST) ([]Task, error) {
	return ml.findWhere(eval(tree.Expression)), nil
}

func eval(exp ast.Expression) func(t Task) bool {
	switch exp.(type) {
	case ast.InfixExpression:
		return evalInfixExp(exp.(ast.InfixExpression))
	case ast.StringLiteral:
		return evalStrLiteralQ(exp.(ast.StringLiteral))
	default:
		return func(t Task) bool { return false }
	}
}

func evalInfixExp(exp ast.InfixExpression) func(t Task) bool {
	switch exp.Operator.Type {
	case token.AND:
		return func(t Task) bool {
			return eval(exp.Left)(t) && eval(exp.Right)(t)
		}
	case token.OR:
		return func(t Task) bool {
			return eval(exp.Left)(t) || eval(exp.Right)(t)
		}
	case token.LIKE:
		return like(exp.Left, exp.Right)
	case token.NLIKE:
		return func(t Task) bool { return !like(exp.Left, exp.Right)(t) }
	case token.EQ:
		return eq(exp.Left, exp.Right)
	case token.NE:
		return func(t Task) bool { return !eq(exp.Left, exp.Right)(t) }
	case token.GT:
		return gt(exp.Left, exp.Right)
	case token.GTE:
		return gte(exp.Left, exp.Right)
	case token.LT:
		return func(t Task) bool { return !gt(exp.Left, exp.Right)(t) }
	case token.LTE:
		return lte(exp.Left, exp.Right)
	default:
		return func(t Task) bool { return false }
	}
}

func evalStrLiteralQ(exp ast.StringLiteral) func(t Task) bool {
	return func(t Task) bool {
		return strings.Contains(t.Title, exp.Value) ||
			strings.Contains(t.Body, exp.Value)
	}
}

func gt(left, right ast.Expression) func(Task) bool {
	return func(t Task) bool {
		switch left.(ast.StringLiteral).Value {
		case "priority":
			return t.Priority > right.(ast.NumberLiteral).Value
		case "created_date":
			fallthrough
		case "createdDate":
			return t.CreatedDate.After(right.(ast.DateLiteral).Value)
		case "completed_date":
			fallthrough
		case "completedDate":
			return t.CompletedDate.After(right.(ast.DateLiteral).Value)
		default:
			return false
		}
	}
}

func gte(left, right ast.Expression) func(Task) bool {
	return func(t Task) bool {
		switch left.(ast.StringLiteral).Value {
		case "priority":
			return t.Priority >= right.(ast.NumberLiteral).Value
		case "created_date":
			fallthrough
		case "createdDate":
			return t.CreatedDate.After(right.(ast.DateLiteral).Value) ||
				t.CreatedDate == right.(ast.DateLiteral).Value
		case "completed_date":
			fallthrough
		case "completedDate":
			return t.CompletedDate.After(right.(ast.DateLiteral).Value) ||
				t.CompletedDate == right.(ast.DateLiteral).Value
		default:
			return false
		}
	}
}

func lte(left, right ast.Expression) func(Task) bool {
	return func(t Task) bool {
		switch left.(ast.StringLiteral).Value {
		case "priority":
			return t.Priority <= right.(ast.NumberLiteral).Value
		case "created_date":
			fallthrough
		case "createdDate":
			return t.CreatedDate.Before(right.(ast.DateLiteral).Value) ||
				t.CreatedDate == right.(ast.DateLiteral).Value
		case "completed_date":
			fallthrough
		case "completedDate":
			return t.CompletedDate.Before(right.(ast.DateLiteral).Value) ||
				t.CompletedDate == right.(ast.DateLiteral).Value
		default:
			return false
		}
	}
}

func eq(left, right ast.Expression) func(Task) bool {
	return func(t Task) bool {
		switch left.(ast.StringLiteral).Value {
		case "title":
			return t.Title == right.(ast.StringLiteral).Value
		case "body":
			return t.Body == right.(ast.StringLiteral).Value
		case "context":
			return t.Context == right.(ast.StringLiteral).Value
		case "priority":
			return t.Priority == right.(ast.NumberLiteral).Value
		case "created_date":
			fallthrough
		case "createdDate":
			return t.CreatedDate == right.(ast.DateLiteral).Value
		case "completed_date":
			fallthrough
		case "completedDate":
			return t.CompletedDate == right.(ast.DateLiteral).Value
		case "completed":
			boolStr := right.(ast.StringLiteral).Value
			completed, err := strconv.ParseBool(boolStr)
			if err != nil {
				return false
			}

			return t.CompletedDate.IsZero() != completed
		default:
			return false
		}
	}
}

func like(left, right ast.Expression) func(Task) bool {
	return func(t Task) bool {
		switch left.(ast.StringLiteral).Value {
		case "title":
			return strings.Contains(t.Title, right.(ast.StringLiteral).Value)
		case "body":
			return strings.Contains(t.Body, right.(ast.StringLiteral).Value)
		case "context":
			return strings.Contains(t.Context, right.(ast.StringLiteral).Value)
		default:
			return false
		}
	}
}
